debruijn/src/SimpleTyped.hs

```diff
@@ -101,4 +101,115 @@ reduce (AppE (LamE t e1) e2)   = subst (singleSub (reduce e2)) (reduce e1)
 -- reduce (AppE (IntE x)    e2)   = error "Type error!"
 reduce (AppE e1 e2) = AppE (reduce e1) (reduce e2)
 
+-----------------------------------------------------------------------
+-- Evaluation to normal form
+
+-- The key take away is that this is mostly boilerplate.
+
+-- Yet another variant of 'Exp'.
+--
+-- How handy that Neutral and Normal start with letter N.
+--
+newtype Norm g t = Norm (GenNorm Norm g t)
+type Neut = GenNeut Norm
+
+-- We define more generic version of 'Norm' and 'Neut',
+-- so we can define
+data GenNeut :: ([Ty] -> Ty -> Type) -> [Ty] -> Ty -> Type where
+ VarN   :: Idx g t               -- variable index
+        -> GenNeut norm g t
+
+ AppN   :: GenNeut norm g (t1 :-> t2)     -- function
+        -> norm      g t1              -- argument
+        -> GenNeut norm g t2
+
+data GenNorm :: ([Ty] -> Ty -> Type) -> [Ty] -> Ty -> Type where
+ IntN   :: Int
+        -> GenNorm norm g IntTy
 
+ LamN   :: Π (t1 :: Ty)           -- type of binder
+        -> norm (t1:g) t2         -- body of abstraction
+        -> GenNorm norm g (t1 :-> t2)
+
+ NeutN :: GenNeut norm g t -> GenNorm norm g t
+
+-- | Like 'reduce' but using 'Norm' as result.
+--
+-- We can remove the equation
+--
+--     reduce (AppE (LamE t e1) e2) = ...
+--
+-- from the definition of 'reduce' - and we won't notice the cheat.
+-- with 'Norm' type-check is unforgiving.
+--
+-- (In fact, I think that 'reduce'  some redexes)
+--
+reduceN :: Exp g t -> Norm g t
+reduceN (IntE x)   = Norm (IntN x)
+reduceN (VarE x)   = Norm (NeutN (VarN x))
+reduceN (LamE t e) = Norm (LamN t (reduceN e))
+reduceN (AppE e1 e2) = case evalW e1 of
+  NeutN e1'  -> Norm (NeutN (AppN (reduceW e1') (reduceN e2)))
+  LamN t e1' ->
+      -- here we have various reduction strategies
+      -- The following is easiest without any additional machinery,
+      -- substitute terms as is. This most likely duplicates work.
+      --
+      --   reduceN (subst (singleSub e2) e1')
+      --
+      -- or we can do
+      --
+      --  reduceN (subst (singleSub (norm2exp (reduceN e2))) (norm2exp (reduceN e1')))
+      --  ^ I think the reduceN after subst is missing from 'reduce' on 'Exp'.
+      --
+      -- or we can define substitution on 'Norm'!
+      --
+      subst (singleSub (reduceN e2)) (reduceN e1')
+
+      -- these three strategies reduce differenty, but
+      -- the types tell that we get a normal form at the end.
+      -- (And the meta-theory tells that it will be the same).
+
+-- again, (almost) the same instance definition as before.
+instance SubstDB Norm where
+   var = Norm . NeutN . VarN
+
+   subst s (Norm (IntN x))         = Norm (IntN x)
+   subst s (Norm (LamN ty e))      = Norm (LamN ty (subst (lift s) e))
+   subst s (Norm (NeutN (VarN x))) = applySub s x
+
+   -- substitution in application causes evaluation,
+   -- i.e. recursive substitutions.
```

Comment:
Awesome! I didn't realize that hereditary substitutions work with this interface too.